src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+                            {
+                                converter = new ConvertToListSyntaxEditor(
+                                    forEachStatement,
+                                    convertingNodes,
+                                    invocationExpression.ArgumentList.Arguments.Single().Expression,
+                                    memberAccessExpression.Expression);
+                                return true;
+                            }
+
+                            break;
+                    }
+
+                    break;
+
+                case SyntaxKind.YieldReturnStatement:
+                    var memberDeclaration = semanticModel.GetEnclosingSymbol(forEachStatement.SpanStart, cancellationToken).DeclaringSyntaxReferences.Single().GetSyntax();
+                    var yieldStatements = memberDeclaration.DescendantNodes().OfType<YieldStatementSyntax>();
+
+                    if (forEachStatement.IsParentKind(SyntaxKind.Block) && forEachStatement.Parent.Parent == memberDeclaration)
+                    {
+                        var statementsOnBlockWithForEach = ((BlockSyntax)forEachStatement.Parent).Statements;
+                        var lastStatement = statementsOnBlockWithForEach.Last();
+                        if (yieldStatements.Count() == 1 && lastStatement == forEachStatement)
+                        {
+                            converter = new YieldReturnConverter(forEachStatement, convertingNodes, (YieldStatementSyntax)statementCannotBeConverted);
+                            return true;
+                        }
+
+                        // foreach()
+                        // {
+                        //   yield return ...;
+                        // }
+                        // yield break;
+                        // end of member
+                        if (yieldStatements.Count() == 2 &&
+                            lastStatement.Kind() == SyntaxKind.YieldBreakStatement &&
+                            statementsOnBlockWithForEach.ElementAt(statementsOnBlockWithForEach.Count - 2) == forEachStatement)
+                        {
+                            // This removes the yield break.
+                            converter = new YieldReturnConverter(forEachStatement, convertingNodes, (YieldStatementSyntax)statementCannotBeConverted, nodeToDelete: lastStatement);
+                            return true;
+                        }
+                    }
+
+                    break;
+            }
+
+            converter = default;
+            return false;
+        }
+
+        protected override SyntaxNode AddLinqUsing(SyntaxNode root)
+        {
+            const string linqNamespaceName = "System.Linq";
+            if (root is CompilationUnitSyntax compilationUnit)
+            {
+                var existingUsings = compilationUnit.Usings;
+                if (!existingUsings.Any(existingUsing => existingUsing.Name.ToString().Equals(linqNamespaceName)))
```

Comment:
what if the user has a top level namespace and has System.Linq inside of that?

Reply:
How about using [SemanticModel.LookupNamespacesAndTypes](http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/SemanticModel.cs,575) API?

---
In reply to: [191058356](https://github.com/dotnet/roslyn/pull/27130#discussion_r191058356) [](ancestors = 191058356)

Reply:
Thank you!

semanticModel.GetUsingNamespacesInScope works

SemanticModel.LookupNamespacesAndTypes returns only namespaces already used in the scope not just available.

---
In reply to: [191886433](https://github.com/dotnet/roslyn/pull/27130#discussion_r191886433) [](ancestors = 191886433,191058356)